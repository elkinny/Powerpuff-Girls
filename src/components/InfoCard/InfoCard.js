import React from 'react';
import PropTypes from 'prop-types';
import Link from 'next/link';
import clsx from 'clsx';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faAngleDoubleRight } from '@fortawesome/free-solid-svg-icons';

import style from './style.module.scss';

const InfoCard = ({
  image, name, picDetails, summary, details, subTitle, link, isColumn,
}) => (
  <div className={clsx(style.description, { [style.row]: !isColumn })}>
    <div>
      <img className={style.picture} src={image} alt={name} />
      {picDetails && (
        <div className={style['pic-details']}>
          {picDetails.map((el) => <span key={el} className={style['pic-detail']}>{el}</span>)}
        </div>
      )}
    </div>
    <div className={style.text}>
      <h1 className={style.header}>
        {name}
        <span className={style.subtitle}>
          {subTitle}
        </span>
      </h1>
      <div className={style.summary} dangerouslySetInnerHTML={{ __html: summary }} />
      {link.url && (
      <Link href={link.url}>
        <a href={link.url} className={style.link}>
          <FontAwesomeIcon icon={faAngleDoubleRight} />
          {link.text}
        </a>
      </Link>
      )}
      {details && (
        <div className={style.details}>
          {details.map((el, i) => <p key={i}>{el}</p>)}
        </div>
      )}
    </div>
  </div>
);

InfoCard.defaultProps = {
  picDetails: [],
  details: [],
  link: {},
  summary: 'No summary',
  isColumn: false,
};

InfoCard.propTypes = {
  image: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  picDetails: PropTypes.arrayOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
  ])),
  summary: PropTypes.string,
  details: PropTypes.array,
  subTitle: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
  ]).isRequired,
  link: PropTypes.shape({
    url: PropTypes.string,
    text: PropTypes.string,
  }),
  isColumn: PropTypes.bool,
};

export default InfoCard;
