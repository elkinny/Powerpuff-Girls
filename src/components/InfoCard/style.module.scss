@import '../../styles/constants';
@import '../../styles/mixins';

.description {
  display: flex;
  max-width: 800px;
  margin: 0 auto 100px;
  flex-direction: column;
  align-items: center;

  &.row {
    @include respond('md') {
      flex-direction: row;
    }
    
    .header {
      font-size: 35px;

      @include respond('lg') {
        font-size: 45px;
      }

      @include respond('md') {
        text-align: left;
      }
    }

    .text {
      @include respond('md') {
        margin-left: 60px;
      }
    }

    .picture {
      height: 400px;

      @include respond('sm') {
        height: 400px;
      }
    }
  }

  .header {
    margin: 20px 0 40px;
    text-align: center;
    font-size: 45px;
    line-height: 50px;

    .subtitle {
      color: $colour-white;
      font-size: 14px;
      margin-left: 10px;

      svg {
        color: $colour-white;
        margin: 0 3px;
        height: 10px;
        margin-top: 0;
        display: inline-block;
      }
    }
  }

  .picture {
    display: block;
    height: 100px;
    box-shadow: 4px 4px 20px #000;
    margin: auto;

    @include respond('sm') {
      height: 200px;
    }

    @include respond('md') {
      height: 400px;
    }
  }

  .pic-details {
    display: flex;
    flex-direction: row;
    justify-content: center;
    flex-wrap: wrap;
    margin-top: 10px;

    .pic-detail {
      display: block;
      padding: 5px 15px;
      background-color: $colour-grey;
      margin: 7px;
      border-radius: 4px;
    }
  }

  .text {
    .summary {
      p {
        text-align: justify;
      }
    }

    .details {
      margin-top: 40px;
      color: $colour-steal;

      p {
        margin: 10px 0;

        svg {
          display: inline-block;
          margin-right: 10px;
          margin-bottom: -3px;
          color: $colour-white;
          height: 20px;
        }
      }
    }

    .link {
      color: $colour-blue;
      font-weight: bold;
      font-size: 20px;
      padding: 6px 12px;
      border-bottom: 1px solid $colour-blue;

      &:hover {
        svg {
          transform: translateX(6px);
        }
      }

      svg {
        height: 20px;
        margin-right: 10px;
        margin-bottom: -3px;
        transition: all 0.5s;
      }
    }
  }
}
