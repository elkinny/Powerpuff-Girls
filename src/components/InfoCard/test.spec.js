import React from 'react';
import { render } from '@testing-library/react';

import InfoCard from './InfoCard';

jest.mock('./style.module.scss', () => ({
  row: 'row',
}));

describe('InfoCard', () => {
  test('renders InfoCard component with full data', () => {
    const { container } = render(<InfoCard
      image="/placeholder.png"
      name="name"
      picDetails={['picDetail1', 'picDetail2']}
      summary="Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."
      details={['detail1', 'detail2']}
      subTitle="subTitle"
      link={{ url: '/', text: 'link' }}
    />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders InfoCard component with no picDetails, details and link, but with isColumn', () => {
    const { container } = render(<InfoCard
      image="/placeholder.png"
      name="name"
      summary="Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."
      subTitle="subTitle"
      isColumn
    />);

    expect(container).toMatchSnapshot();
  });
});
