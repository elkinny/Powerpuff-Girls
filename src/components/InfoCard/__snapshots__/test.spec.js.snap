// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoCard renders InfoCard component with full data 1`] = `
<div
  class="row"
>
  <div>
    <img
      alt="name"
      src="/placeholder.png"
    />
    <div>
      <span>
        picDetail1
      </span>
      <span>
        picDetail2
      </span>
    </div>
  </div>
  <div>
    <h1>
      name
      <span>
        subTitle
      </span>
    </h1>
    <div>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
    </div>
    <a
      href="/"
    >
      FontAwesomeIcon
      link
    </a>
    <div>
      <p>
        detail1
      </p>
      <p>
        detail2
      </p>
    </div>
  </div>
</div>
`;

exports[`InfoCard renders InfoCard component with no picDetails, details and link, but with isColumn 1`] = `
<div>
  <div
    class=""
  >
    <div>
      <img
        alt="name"
        src="/placeholder.png"
      />
      <div />
    </div>
    <div>
      <h1>
        name
        <span>
          subTitle
        </span>
      </h1>
      <div>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
      </div>
      <div />
    </div>
  </div>
</div>
`;
